src/ConnectedMode.UnitTests/Migration/ConnectedModeMigrationTests.cs

```diff
@@ -189,21 +192,65 @@ public async Task Migrate_HasProgressListener_ReportsProgress()
             progressMessages.Should().NotBeEmpty();
         }
 
+        [TestMethod]
+        [DataRow(true, false)]
+        [DataRow(false, true)]
+        public async Task Migrate_ConnectToSonarQubeIfNeeded_DisconnectedNotCalled(bool isConnectedToNewConnectedMode, bool expectedResult)
+        {
+            var sonarQubeService = CreateSonarQubeService(isConnectedToNewConnectedMode);
+
+            var testSubject = CreateTestSubject(sonarQubeService: sonarQubeService.Object);
+            await testSubject.MigrateAsync(AnyBoundProject, null, CancellationToken.None);
+
+            sonarQubeService.Verify(x => x.ConnectAsync(It.IsAny<ConnectionInformation>(), CancellationToken.None), expectedResult ? Times.Once : Times.Never);
+            sonarQubeService.Verify(x => x.Disconnect(), Times.Never);
+        }
+
+        [TestMethod]
+        [DataRow(true, false)]
+        [DataRow(false, true)]
+        public async Task Migrate_DisconnectToSonarQubeIfNeeded_NonCritical_HandledAndThrown(bool isConnectedToNewConnectedMode, bool expectedResult)
```

Comment:
Same question about the second argument here - do we need both?

`isAlreadyConnected` might be better than `isConnectedToNewConnectedMode`.